### PR TITLE
Review logical decoding client tracking of LSNs.

### LINKED ISSUE #500: [online] On setting endpos follow mode switches as replay => catchup => replay and gets stuck
Here is the backtrace of stuck transform process.
```
#0  0x00007f13b6091f7a in __GI___select (nfds=nfds@entry=13, readfds=readfds@entry=0x7ffe6149c3d0, writefds=writefds@entry=0x0, exceptfds=exceptfds@entry=0x0, timeout=timeout@entry=0x7ffe6149c3c0) at ../sysdeps/unix/sysv/linux/select.c:41
#1  0x0000555c90c8f2c7 in read_from_stream (stream=0x555c91c7a5e0, context=context@entry=0x7ffe6149e4b0) at file_utils.c:445
#2  0x0000555c90ca1894 in stream_transform_stream (specs=specs@entry=0x7ffe6149ea20) at ld_transform.c:102
#3  0x0000555c90ca2a1a in follow_start_transform (specs=0x7ffe6149ea20) at follow.c:667
#4  0x0000555c90ca21fd in follow_start_subprocess (specs=specs@entry=0x7ffe6149ea20, subprocess=subprocess@entry=0x7ffe615a7760) at follow.c:776
#5  0x0000555c90ca2bc3 in follow_start_subprocess (specs=0x7ffe6149ea20, subprocess=0x7ffe615a7760) at follow.c:761
#6  0x0000555c90ca33eb in followDB (copySpecs=<optimized out>, streamSpecs=0x7ffe6149ea20) at follow.c:542
#7  followDB (copySpecs=<optimized out>, streamSpecs=0x7ffe6149ea20) at follow.c:476
#8  0x0000555c90ca3672 in follow_main_loop (copySpecs=copySpecs@entry=0x7ffe615aa1c0, streamSpecs=streamSpecs@entry=0x7ffe6149ea20) at follow.c:286
#9  0x0000555c90c814f6 in start_follow_process (pid=<synthetic pointer>, streamSpecs=0x7ffe6149ea20, copySpecs=0x7ffe615aa1c0) at cli_clone_follow.c:677
#10 clone_and_follow (copySpecs=0x7ffe615aa1c0) at cli_clone_follow.c:269
#11 cli_clone (argc=<optimized out>, argv=<optimized out>) at cli_clone_follow.c:142
#12 0x0000555c90ccf10c in commandline_run (command=<optimized out>, argc=0, argv=0x7ffe616bdc10) at /home/shdhama/pg/pgcopydb/src/bin/pgcopydb/../lib/subcommands.c/commandline.c:71
#13 0x0000555c90c808f6 in main (argc=15, argv=0x7ffe616bdb98) at main.c:167
```

I'm able to reproduce this without any activity on source.
Here are complete logs:
[date-23-10-16--time-15-18-14.log](https://github.com/dimitri/pgcopydb/files/12919180/date-23-10-16--time-15-18-14.log)

### PR DESCRIPTION
**Pull request by dimitri**
When replaying from file, ensure that the last transaction of a file is commited with synchronous_commit on, allowing better tracking of replication progress.

Also, improve our way to keep track of the previousLSN and log when we reach endpos (KEEPALIVE, COMMIT, ENDPOS) to help get a sense of what's happening.

By setting previousLSN to the proper value even when reaching endpos at a KEEPALIVE message that we skip, we allow the overall system to exit properly in case when there is no activity on the source server.

Fixes #500 
See #427 
See #493

### CODE REVIEW

File: src/bin/pgcopydb/ld_apply.c [hunk 1]
```diff
@@ -394,16 +385,16 @@ stream_apply_sync_sentinel(StreamApplyContext *context)
...
 	{
 		durableLSN = context->previousLSN;
 	}
 	else
 	{
 		if (!stream_apply_find_durable_lsn(context, &durableLSN))
 		{
-			log_debug("Skipping sentinel replay_lsn: "
-					  "failed to find a durable LSN matching current flushLSN");
+			log_warn("Skipping sentinel replay_lsn udpate: "
```
Comment (arajkumar) [inline src/bin/pgcopydb/ld_apply.c:396]:
```suggestion
			log_warn("Skipping sentinel replay_lsn update: "
```

File: src/bin/pgcopydb/ld_apply.c [hunk 2]
```diff
@@ -1125,15 +1163,41 @@ setupReplicationOrigin(StreamApplyContext *context, bool logSQL)
...
+	if (context->previousLSN == InvalidXLogRecPtr)
+	{
+		log_info("Setting up previous LSN from "
+				 "replication origin \"%s\" progress at %X/%X",
+				 nodeName,
+				 LSN_FORMAT_ARGS(originLSN));
+
+		context->previousLSN = originLSN;
+	}
+	else if (context->previousLSN != originLSN)
```
Comment (arajkumar) [inline src/bin/pgcopydb/ld_apply.c:1189]: I'm wondering when this condition would hit? Wouldn't that `context->previousLSN` always `InvalidXLogRecPtr` ?

**Review by arajkumar — APPROVED** [review]
lgtm

**Comment by arajkumar** [comment]
I understand that the PR doesn't handle the following scenario, but just adding my observation while testing this PR.

When I set endpos, sometimes `transform` blocks while writing to output stream.

```
➜  pgcopydb git:(fix/previous_lsn_tracking) pgrep -a pgcopydb
547254 pgcopydb follow --fail-fast --plugin wal2json --restart -vv
547275 pgcopydb: follow prefetch
547276 pgcopydb: follow transform
```

## tranform
```
#0  0x00007f79ac714a77 in __GI___libc_write (fd=9, buf=0x55be28408e60, nbytes=4096) at ../sysdeps/unix/sysv/linux/write.c:26
#1  0x00007f79ac68aeed in _IO_new_file_write (f=0x55be283d9d00, data=0x55be28408e60, n=4096) at ./libio/fileops.c:1180
#2  0x00007f79ac68c9e1 in new_do_write (to_do=4096,
    data=0x55be28408e60 "l,\"2016-01-24 06:39:50+00\",\"3890\",\"0.700000\",\"1431\",\"0\",null,\"2016-01-24 06:39:50+00\",\"3886\",\"0.800000\",\"1431\",\"0\",null,\"2016-01-24 06:39:50+00\",\"7993\",\"0.800000\",\"1431\",\"5\",null,\"2016-01-24 06:40:00+"..., fp=0x55be283d9d00) at ./libio/libioP.h:947
#3  _IO_new_do_write (to_do=4096,
    data=0x55be28408e60 "l,\"2016-01-24 06:39:50+00\",\"3890\",\"0.700000\",\"1431\",\"0\",null,\"2016-01-24 06:39:50+00\",\"3886\",\"0.800000\",\"1431\",\"0\",null,\"2016-01-24 06:39:50+00\",\"7993\",\"0.800000\",\"1431\",\"5\",null,\"2016-01-24 06:40:00+"..., fp=0x55be283d9d00) at ./libio/fileops.c:425
#4  _IO_new_do_write (fp=0x55be283d9d00,
    data=0x55be28408e60 "l,\"2016-01-24 06:39:50+00\",\"3890\",\"0.700000\",\"1431\",\"0\",null,\"2016-01-24 06:39:50+00\",\"3886\",\"0.800000\",\"1431\",\"0\",null,\"2016-01-24 06:39:50+00\",\"7993\",\"0.800000\",\"1431\",\"5\",null,\"2016-01-24 06:40:00+"..., to_do=4096) at ./libio/fileops.c:422
#5  0x00007f79ac68b6d5 in _IO_new_file_xsputn (n=1024, data=<optimized out>, f=<optimized out>) at ./libio/libioP.h:947
#6  _IO_new_file_xsputn (f=0x55be283d9d00, data=<optimized out>, n=1024) at ./libio/fileops.c:1196
#7  0x00007f79ac67ffd7 in __GI__IO_fwrite (buf=0x7ffdb5577460, size=size@entry=1, count=count@entry=1024, fp=0x55be283d9d00) at ./libio/libioP.h:947
#8  0x000055be277f6ea1 in flushbuffer (target=target@entry=0x7ffdb5577430) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/../lib/pg/snprintf.c:311
#9  0x000055be277f6f18 in dostr (target=<optimized out>, slen=1511,
    str=0x55be29320180 "\",\"7970\",\"0.700000\",\"3126\",\"2\",null,\"2016-01-24 06:41:10+00\",\"3868\",\"0.800000\",\"3126\",\"4\",null,\"2016-01-24 06:41:10+00\",\"3874\",\"0.700000\",\"3126\",\"2\",null,\"2016-01-24 06:41:10+00\",\"3870\",\"0.800000\",\"31"...) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/../lib/pg/snprintf.c:1385
#10 dostr (
    str=str@entry=0x55be29316d90 "[\"2016-01-24 06:31:20+00\",\"3886\",\"0.800000\",\"635\",\"3\",null,\"2016-01-24 06:31:20+00\",\"3890\",\"0.800000\",\"635\",\"2\",null,\"2016-01-24 06:31:20+00\",\"7993\",\"0.800000\",\"635\",\"4\",null,\"2016-01-24 06:31:30+00\","..., slen=<optimized out>, target=target@entry=0x7ffdb5577430)
    at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/../lib/pg/snprintf.c:1360
#11 0x000055be277f9b46 in dopr (target=target@entry=0x7ffdb5577430, format=0x55be2781da55 ";\n", args=<optimized out>)
    at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/../lib/pg/snprintf.c:427
#12 0x000055be277fafbc in pg_vfprintf (stream=<optimized out>, fmt=<optimized out>, args=<optimized out>)
    at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/../lib/pg/snprintf.c:258
#13 0x000055be277d0fbf in fformat (stream=<optimized out>, fmt=<optimized out>) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/file_utils.c:1188
#14 0x000055be277de06b in stream_write_insert (insert=0x55be29327af8, out=0x55be283d9d00) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_transform.c:2056
#15 stream_write_transaction (txn=0x7ffdb5682c80, out=0x55be283d9d00) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_transform.c:1748
#16 stream_write_message (out=0x55be283d9d00, msg=msg@entry=0x7ffdb5682c78) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_transform.c:1605
#17 0x000055be277df362 in stream_transform_write_message (privateContext=privateContext@entry=0x7ffdb5681038, currentMsgIndex=currentMsgIndex@entry=0x7ffdb5577c78)
    at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_transform.c:336
#18 0x000055be277e56f6 in stream_transform_line (ctx=0x7ffdb5577c70,
    line=0x55be292f1660 "{\"action\":\"C\",\"xid\":\"7438006\",\"lsn\":\"148/488F5FF8\",\"timestamp\":\"2023-10-19 06:11:23.1918+0000\",\"message\":{\"action\":\"C\",\"xid\":7438006}}", stop=0x7ffdb5577b87) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_transform.c:267
#19 0x000055be277cfdfd in read_from_stream (stream=<optimized out>, context=0x7ffdb5577c80) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/file_utils.c:643
#20 0x000055be277e5171 in stream_transform_stream (specs=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_transform.c:123
#21 0x000055be277ce800 in follow_start_transform (specs=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:673
#22 follow_start_transform (specs=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:642
#23 0x000055be277d372a in follow_start_subprocess (specs=specs@entry=0x7ffdb55782a0, subprocess=subprocess@entry=0x7ffdb5680fe0)
    at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:782
#24 0x000055be277d3ce3 in followDB (streamSpecs=0x7ffdb55782a0, copySpecs=<optimized out>) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:542
#25 0x000055be278075a0 in followDB (copySpecs=<optimized out>, streamSpecs=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:493
#26 follow_main_loop.part.0.isra.0 (streamSpecs=0x7ffdb55782a0, copySpecs=<optimized out>) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:286
```

## prefetch
```
#0  0x00007f79ac714a77 in __GI___libc_write (fd=7, buf=0x55be292f0050, nbytes=791) at ../sysdeps/unix/sysv/linux/write.c:26
#1  0x00007f79ac68aeed in _IO_new_file_write (f=0x55be283da730, data=0x55be292f0050, n=791) at ./libio/fileops.c:1180
#2  0x00007f79ac68c9e1 in new_do_write (to_do=791,
    data=0x55be292f0050 "{\"action\":\"I\",\"xid\":\"7438004\",\"lsn\":\"148/48888E58\",\"timestamp\":\"2023-10-19 06:11:23.3124+0000\",\"message\":{\"action\":\"I\",\"xid\":7438004,\"schema\":\"public\",\"table\":\"readings\",\"columns\":[{\"name\":\"time\",\"typ"..., fp=0x55be283da730) at ./libio/libioP.h:947
#3  _IO_new_do_write (to_do=791,
    data=0x55be292f0050 "{\"action\":\"I\",\"xid\":\"7438004\",\"lsn\":\"148/48888E58\",\"timestamp\":\"2023-10-19 06:11:23.3124+0000\",\"message\":{\"action\":\"I\",\"xid\":7438004,\"schema\":\"public\",\"table\":\"readings\",\"columns\":[{\"name\":\"time\",\"typ"..., fp=0x55be283da730) at ./libio/fileops.c:425
#4  _IO_new_do_write (fp=0x55be283da730,
    data=0x55be292f0050 "{\"action\":\"I\",\"xid\":\"7438004\",\"lsn\":\"148/48888E58\",\"timestamp\":\"2023-10-19 06:11:23.3124+0000\",\"message\":{\"action\":\"I\",\"xid\":7438004,\"schema\":\"public\",\"table\":\"readings\",\"columns\":[{\"name\":\"time\",\"typ"..., to_do=791) at ./libio/fileops.c:422
#5  0x00007f79ac68b6d5 in _IO_new_file_xsputn (n=791, data=<optimized out>, f=<optimized out>) at ./libio/libioP.h:947
#6  _IO_new_file_xsputn (f=0x55be283da730, data=<optimized out>, n=791) at ./libio/fileops.c:1196
#7  0x00007f79ac67ffd7 in __GI__IO_fwrite (buf=buf@entry=0x55be292ee550, size=size@entry=1, count=count@entry=791, fp=fp@entry=0x55be283da730) at ./libio/libioP.h:947
#8  0x000055be277dc529 in write_to_stream (size=<optimized out>,
    buffer=0x55be292ee550 "{\"action\":\"I\",\"xid\":\"7438004\",\"lsn\":\"148/48888E58\",\"timestamp\":\"2023-10-19 06:11:23.3124+0000\",\"message\":{\"action\":\"I\",\"xid\":7438004,\"schema\":\"public\",\"table\":\"readings\",\"columns\":[{\"name\":\"time\",\"typ"..., stream=0x55be283da730)
    at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/file_utils.c:392
#9  stream_write_json (context=context@entry=0x7ffdb5370180, previous=previous@entry=false) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_stream.c:798
#10 0x000055be277dcbc2 in streamWrite (context=0x7ffdb5370180) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_stream.c:688
#11 0x000055be277d95c9 in pgsql_stream_logical (context=0x7ffdb5370180, client=0x7ffdb5470e10) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/pgsql.c:4189
#12 startLogicalStreaming (specs=specs@entry=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/ld_stream.c:446
#13 0x000055be277da728 in follow_start_prefetch (specs=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:614
#14 follow_start_prefetch (specs=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:594
#15 0x000055be277d372a in follow_start_subprocess (specs=specs@entry=0x7ffdb55782a0, subprocess=subprocess@entry=0x7ffdb5680fc0)
    at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:782
#16 0x000055be277d3cc8 in followDB (streamSpecs=0x7ffdb55782a0, copySpecs=<optimized out>) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:536
#17 0x000055be278075a0 in followDB (copySpecs=<optimized out>, streamSpecs=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:493
#18 follow_main_loop.part.0.isra.0 (streamSpecs=0x7ffdb55782a0, copySpecs=<optimized out>) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:286
#19 0x000055be277b4dda in follow_main_loop (copySpecs=0x7ffdb5683a40, streamSpecs=0x7ffdb55782a0) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/follow.c:262
#20 cli_follow (argc=<optimized out>, argv=<optimized out>) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/cli_clone_follow.c:425
#21 0x000055be277b2d01 in commandline_run (argv=0x7ffdb5797480, argc=0, command=0x7ffdb5796d90)
    at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/../lib/subcommands.c/commandline.c:71
#22 main (argc=<optimized out>, argv=<optimized out>) at /home/ubuntu/dimitri/pgcopydb/src/bin/pgcopydb/main.c:167
```

File: src/bin/pgcopydb/ld_apply.c (hunk 2 above)
Comment (dimitri) [reply src/bin/pgcopydb/ld_apply.c:1189]: It could have been initialized from reading the value in the sentinel at startup, when re-launching the pgcopydb command from a previous run. I have added a comment explaining that, thanks.

File: src/bin/pgcopydb/ld_apply.c (hunk 1 above)
Comment (dimitri) [reply src/bin/pgcopydb/ld_apply.c:396]: Applied locally, will be in the next push. Thanks!

**Comment by dimitri** [comment]
> I understand that the PR doesn't handle the following scenario, but just adding my observation while testing this PR.

Yeah I'm just trying to get my head around the problem space, fixing the issues I can reproduce easily on my test environment for now. This PR fixes an actual issue that I can't reproduce anymore, so that's good, but like you I believe some more important issues need taken care of. I am currently trying to diagnose another issue that manifests itself as a Broken PIPE at write time from ld_stream. @shubhamdhama also found an issue with endpos being set in the middle of a transaction in such a way that then we never reach it: replay_lsn < endpos; and that also causes an infinite loop.

In short, more to come, it seems like we need to do many smaller things before we can attack the main problems.
